artifacts/definitions/Windows/Remediation/Sinkhole.yaml

```diff
@@ -0,0 +1,187 @@
+      LET existing = SELECT
+            parse_string_with_regex(
+            string=Line,
+            regex=[
+                "^\\s+(?P<Resolution>[^\\s]+)\\s+" +
+                "(?P<Hostname>[^\\s]+)\\s*\\S*$"
+            ]) as Record,
+            Line
+        FROM parse_lines(filename=HostsFile)
+        WHERE 
+            Line 
+            AND NOT Line =~ '^#'
+
+      -- Parse a URL to get domain name.
+      LET get_domain(URL) = parse_string_with_regex(
+           string=URL, regex='^https?://(?P<Domain>[^:/]+)').Domain
+
+      -- extract Velociraptor config for policy
+      LET extracted_config <= SELECT * FROM foreach(
+          row=config.server_urls,
+            query={
+                SELECT get_domain(URL=_value) AS Domain
+                FROM scope()
+            })
+
+      -- Set existing entries to sinkholed values
+      LET find_modline = SELECT * FROM foreach(row=changes,
+            query={
+                SELECT
+                    format(format='\t%v\t\t%v\t\t# %v',
+                    args=[Sinkhole,Domain,Description]) as Line,
+                    Domain,
+                    'modification' as Type
+                FROM existing
+                WHERE
+                    Record.Hostname = Domain
+                    AND NOT Domain in extracted_config.Domain
+                GROUP BY Line
+            })
+
+      -- Add new hostsfile entries
+      LET find_newline = SELECT * FROM foreach(row=changes,
+            query={
+                SELECT 
+                    format(format='\t%v\t\t%v\t\t# %v',
+                        args=[Sinkhole,Domain,Description]) as Line,
+                    Domain,
+                    'new entry' as Type
+                FROM scope()
+                WHERE
+                    NOT Domain in find_modline.Domain
+                    AND NOT Domain in extracted_config.Domain
+            })
+
+      -- Determine which lines should stay the same
+      LET find_line= SELECT 
+                Line,
+                Record.Hostname as Domain,
+                'old entry' as Type
+            FROM existing
+            WHERE 
+                NOT Domain in find_modline.Domain
+                AND NOT Domain in find_newline.Domain
+
+      -- Add all lines to staging object
+      LET build_lines = SELECT Line FROM chain(
+            a=find_modline,
+            b=find_newline,
+            c=find_line
+      )
+      
+      -- Join lines from staging object
+      LET HostsData = join(array=build_lines.Line,sep='\r\n')
+
+
+      -- Force start of backup or restore if applicable
+      LET backup_restore <= if(condition= RestoreBackup,
+                then= if(condition= check_backup,
+                        then= restore,
+                        else= log(message='Can not restore hosts file as backup does not exist.')),
+                else= if(condition= check_backup,
+                        then={
+                            SELECT * FROM chain(
```

Comment:
If this artifact is collected twice it does **not** overwrite the backup of the original hosts file right? This is the correct thing we just need to document it better.

Reply:
Yes - this is correct each application it calls the restore VQL.

If there is a rollback tickbox it will call the restore and delete the backup and not process anything else. 
If there is no rollback it will restore the backup and apply the new table to it or if there is no backup just copy it and apply the table.